lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -3460,6 +3465,37 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
   return nullptr;
 }
 
+SPIRVValue *LLVMToSPIRVBase::transFenceInst(FenceInst *FI,
+                                            SPIRVBasicBlock *BB) {
+  SPIRVWord MemorySemantics = static_cast<SPIRVWord>(FI->getOrdering());
+  // Fence ordering may only be Acquire, Release, AcquireRelease, or
+  // SequentiallyConsistent
+  switch (MemorySemantics) {
+  case 4:
+    MemorySemantics = 0x2;
+    break;
+  case 5:
+    MemorySemantics = 0x4;
+    break;
+  case 6:
+    MemorySemantics = 0x8;
+    break;
+  case 7:
+    MemorySemantics = 0x10;
+    break;
+  default:
+    assert(false && "Unexpected fence ordering");
+    MemorySemantics = SPIRVWORD_MAX;
+    break;
+  }
+
+  Module *M = FI->getParent()->getModule();
+  SPIRVValue *RetScope = transConstant(getUInt32(M, 0));
```

Comment:
Should we call `FI->getSyncScopeID()` and map the return value to a suitable SPIR-V Scope?

Reply:
Probably I am wrong, but llvm.fence instruction doesn't contain information about SPIR-V Scope (workgroup/subgroup/etc).
SyncScope can only be SingleThread or System https://llvm.org/doxygen/LLVMContext_8h_source.html#l00046

Reply:
Yeah I'm not very confident about a suitable mapping either actually...  But we should make it more obvious that we're assuming `ScopeCrossDevice` then:
```suggestion
  // Treat all llvm.fence instructions as having CrossDevice scope:
  SPIRVValue *RetScope = transConstant(getUInt32(M, ScopeCrossDevice));
```